### PR TITLE
PLG-82: Track customer changes

### PR DESCRIPTION
**Pull request by nedick007**
Reworking customer events to correspond to the new api specification

### CODE REVIEW

**Comment by nedick007** [comment]
@adamof ready for review

File: Controller/Adminhtml/Import/Ajax.php [hunk 1]
```diff
@@ -62,8 +66,17 @@ public function execute()
 
             switch ($importType) {
                 case 'customers':
-                    $client->customerBatch($this->import->customerData->getCustomers($storeId, $chunkId));
-                    $result['success'] = 'customerBatch';
+                    $serializedCustomers = [];
```
Comment (adamof) [inline Controller/Adminhtml/Import/Ajax.php]: are you really going to copy and paste this code 4 times?

File: Helper/CustomerSerializer.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Metrilo\Analytics\Helper;
+
+class CustomerSerializer extends \Magento\Framework\App\Helper\AbstractHelper
+{
+    public function __construct(
+        \Magento\Newsletter\Model\Subscriber $subscriberModel,
+        \Magento\Framework\App\RequestInterface $request
```
Comment (adamof) [inline Helper/CustomerSerializer.php]: What would you need the request for??

File: Helper/Data.php [hunk 3]
```diff
@@ -31,26 +31,28 @@ class Data extends \Magento\Framework\App\Helper\AbstractHelper
...
-        \Metrilo\Analytics\Helper\OrderSerializer $orderSerializer,
-        \Metrilo\Analytics\Helper\AdminStoreResolver $resolver,
-        \Magento\Store\Model\StoreManagerInterface $storeManager,
-        \Magento\Framework\App\ProductMetadata $metaData,
-        \Magento\Framework\Module\ModuleListInterface $moduleList
+        \Magento\Customer\Model\Session                    $session,
+        \Psr\Log\LoggerInterface                           $logger,
+        \Magento\Framework\Json\Helper\Data                $jsonHelper,
+        \Metrilo\Analytics\Helper\Client                   $clientHelper,
+        \Metrilo\Analytics\Helper\CustomerSerializer       $customerSerializer,
```
Comment (adamof) [inline Helper/Data.php]: why???

File: Observer/CustomerAdminUpdate.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    }
+    
+    public function execute(Observer $observer)
+    {
+        try {
+            $customer     = $observer->getEvent()->getCustomer();
+            $origCustomer = $this->customerRepository->getById($customer->getId());
+            $storeId      = $this->helper->getStoreId();
+            
+            if (!trim($customer->getEmail())) {
```
Comment (adamof) [inline Observer/CustomerAdminUpdate.php]: you have copied this code in 4 new places 
https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

File: Observer/CustomerAdminUpdate.php [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                $this->helper->logError('Customer with id = '. $customer->getId(). '  have no email address!');
+                return;
+            }
+            
+            // Create api call only if there is difference between original and saved after customer data.
+            if ($customer->getEmail() == $origCustomer->getEmail() &&
```
Comment (adamof) [inline Observer/CustomerAdminUpdate.php]: does this even work, given the fact that you are in a **after** save hook? Aren't the DB and memory copies the same?

File: Observer/CustomerAdminUpdate.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                $customer->getFirstname() == $origCustomer->getFirstname() &&
+                $customer->getLastname() == $origCustomer->getLastname()) {
+                $hasChanges = false;
+            } else {
+                $hasChanges = true;
+            }
+            
+            if ($hasChanges) {
+                $client             = $this->apiClient->getClient($storeId);
```
Comment (adamof) [inline Observer/CustomerAdminUpdate.php]: same DRY question about this code, too

File: etc/frontend/events.xml [hunk 7]
```diff
@@ -7,11 +7,11 @@
...
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Event/etc/events.xsd">
-    <event name="customer_login">
-        <observer name="metrilo_event_login" instance="Metrilo\Analytics\Observer\CustomerLogin" />
-    </event>
     <event name="customer_account_edited">
-    	<observer name="metrilo_event_account_update" instance="Metrilo\Analytics\Observer\CustomerUpdate" />
+        <observer name="metrilo_event_customer_update" instance="Metrilo\Analytics\Observer\CustomerUpdate" />
+    </event>
+    <event name="customer_register_success">
+        <observer name="metrilo_event_customer_create" instance="Metrilo\Analytics\Observer\CustomerCreate" />
```
Comment (adamof) [inline etc/frontend/events.xml]: Why do you insist on having all the observers in different files?

File: Helper/CustomerSerializer.php (hunk 2 above)
Comment (nedick007) [reply Helper/CustomerSerializer.php]: This is redundant dependency injection. Removed.

**Comment by nedick007** [comment]
@adamof ready for review

File: Observer/CustomerAdminUpdate.php (hunk 5 above)
Comment (adamof) [reply Observer/CustomerAdminUpdate.php]: Lets discuss in person

File: Observer/Customer.php [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        $this->apiClient          = $apiClient;
+        $this->customerSerializer = $customerSerializer;
+        $this->customerRepository = $customerRepository;
+    }
+    
+    public function execute(Observer $observer)
+    {
+        try {
+            $storeId    = $this->helper->getStoreId();
+            $client     = $this->apiClient->getClient($storeId);
```
Comment (adamof) [inline Observer/Customer.php]: why do you initialize the client so early? What if you don't send the event due to no changes or no email?

File: Observer/Customer.php [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                    $email      = $subscriber->getEmail();
+                    $hasChanges = $subscriber->isStatusChanged();
+                    break;
+                case 'customer_account_edited':
+                    $email    = $observer->getEvent()->getEmail();
+                    $customer = $this->customerRepository->get($email);
+                    break;
+                case 'customer_register_success':
+                    $customer = $observer->getEvent()->getCustomer();
+                    $email    = $customer->getEmail();
```
Comment (adamof) [inline Observer/Customer.php]: Why don't you get the email after the case statement with just `$customer->getEmail()`?

File: Observer/Customer.php [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+                    break;
+                default:
+                    break;
+            }
+            
+            if (!trim($email)) {
+                $this->helper->logError('Customer with id = '. $customer->getId(). '  have no email address!');
```
Comment (adamof) [inline Observer/Customer.php]: has*

**Comment by nedick007** [comment]
@adamof ready for review

File: Block/System/Config/Button/Import.php [hunk 11]
```diff
@@ -114,14 +116,24 @@ public function getImport()
         return $this->import;
     }
 
+    /**
+ * Order Import model
+ *
+ * @return \Metrilo\Analytics\Helper\OrderSerializer
```
Comment (adamof) [inline Block/System/Config/Button/Import.php]: Is this accurate?

File: Block/System/Config/Button/Import.php [hunk 12]
```diff
@@ -114,14 +116,24 @@ public function getImport()
...
+ */
+    public function getOrderData()
```
Comment (adamof) [inline Block/System/Config/Button/Import.php]: Why do you need these methods? Can't you get the `orderData` by using `->` operator?

File: Controller/Adminhtml/Import/Ajax.php [hunk 13]
```diff
@@ -62,29 +76,20 @@ public function execute()
...
-                        $serializedOrders[] = $this->orderSerializer->serializeOrder($order);
-                    }
-                    if(!empty($serializedOrders)) {
-                        $client->orderBatch($serializedOrders);
-                        $result['success'] = 'orderBatch';
-                    } else {
-                        $result['success'] = 'empty orderBatch';
-                    }
+                    $serializedOrders  = $this->serializeRecords($this->orderData->getOrders($storeId, $chunkId), $this->orderSerializer);
+                    $result['success'] = $client->customerBatch($serializedOrders);
```
Comment (adamof) [inline Controller/Adminhtml/Import/Ajax.php]: Why are you sending the serialized orders to the customers endpoint?

File: Observer/Customer.php [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        \Metrilo\Analytics\Helper\CustomerSerializer      $customerSerializer,
+        \Magento\Customer\Api\CustomerRepositoryInterface $customerRepository
+    ) {
+        $this->helper             = $helper;
+        $this->apiClient          = $apiClient;
+        $this->customerSerializer = $customerSerializer;
+        $this->customerRepository = $customerRepository;
+    }
+    
+    public function execute(Observer $observer)
```
Comment (adamof) [inline Observer/Customer.php]: Please refactor by defining methods to make understanding this easier

File: Observer/Customer.php [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+                        $customer->getFirstname() == $originalCustomer->getFirstname() &&
+                        $customer->getLastname() == $originalCustomer->getLastname()) {
+                        $hasChanges = false;
+                    } else {
+                        $hasChanges = true;
+                    }
+                    break;
+                case 'newsletter_subscriber_save_after':
+                    $subscriber = $observer->getEvent()->getSubscriber();
+                    $customer   = $this->customerRepository->getById($subscriber->getCustomerId());
```
Comment (adamof) [inline Observer/Customer.php]: you don't have to make the query to fetch the customer if `hasChanges` will be false, right?

File: Observer/Customer.php [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+                    $customer = $this->customerRepository->get($observer->getEvent()->getEmail());
+                    break;
+                case 'customer_register_success':
+                    $customer = $observer->getEvent()->getCustomer();
+                    break;
+                default:
+                    break;
+            }
+            
+            if (!trim($customer->getEmail())) {
```
Comment (adamof) [inline Observer/Customer.php]: no need to do this if `hasChanges` is `false`

File: Observer/Customer.php (hunk 15 above)
Comment (nedick007) [reply Observer/Customer.php]: Customer observer is refactored.

File: Observer/Customer.php (hunk 16 above)
Comment (nedick007) [reply Observer/Customer.php]: Fixed with customer observer refactoring.

File: Observer/Customer.php (hunk 14 above)
Comment (nedick007) [reply Observer/Customer.php]: Customer observer is refactored.

**Comment by nedick007** [comment]
@adamof Ready for review.

**Pull request merged**